### ISSUE TITLE
Inconsistent Behaviour of the plugin

### ISSUE THREAD

**Issue by aswinsegu** [apigee/istio-mixer-adapter#187]
Upon installing the mixer plugin, the behaviour was as expected
<img width="1077" alt="screen shot 2018-08-24 at 18 46 12" src="https://user-images.githubusercontent.com/10217586/44586718-553d9180-a7ce-11e8-9513-9fb53590942b.png">

After sometime, used to provide correct response even though the consumer key is not passed
<img width="1076" alt="screen shot 2018-08-24 at 18 46 50" src="https://user-images.githubusercontent.com/10217586/44586769-7bfbc800-a7ce-11e8-829f-c187eedd6906.png">

And again, even though the consumer key was properly sent, it failed with PERMISSION_DENIED error
<img width="1087" alt="screen shot 2018-08-24 at 18 47 04" src="https://user-images.githubusercontent.com/10217586/44586806-9cc41d80-a7ce-11e8-9374-6e447b9302af.png">

Could not set any consistent behaviour though.
Using new version of the plugin and K8s 1.9.3

**Comment by theganyo** [comment]
This deserves investigation. It is possible that the behavior you're seeing is due to a cache?

**Comment by jaredhwilliams** [comment]
I'm seeing this inconsistent behavior as well with the verify api key functionality. I run several curl commands in a row, using a valid API key, some succeed, others fail with PERMISSION_DENIED.

**Comment by theganyo** [comment]
I'm investigating this now. If there's any way you can recreate this behavior with debug logging on for the adapter, that would be extremely helpful.

**Comment by theganyo** [comment]
Also, if you could provide your platforms and whether you ran istio-demo.yaml or istio-demo-auth.yaml, that would be helpful. Thanks.

**Comment by jaredhwilliams** [comment]
Hi Scott, below are some details about my setup. I deployed the helloworld sample (used istio-demo.yaml) from my local mac, to a GKE cluster.

Istio Mixer Adapter version:
istio-mixer-adapter_1.0.0_macOS_64-bit

K8s version:
kubectl version --short
Client Version: v1.9.7
Server Version: v1.10.5-gke.4

I'll look into getting a debug log.

**Comment by jaredhwilliams** [comment]
What I've found during my troubleshooting is what appears to be "inconsistent" behavior is really due to the fact that applying the rule takes time to get propagated to all instances (in this case helloworld v1 and helloworld vs). This means that after you apply the rule, there will be a short period of time where the rule applies to one version but has not been applied to the other version. Thus, as the load balancer sends your request back and forth to both versions during that time, it will appear that it works sometimes (hitting v1) but not work other times (hitting v2). After a short time, after the rule has propagated to both, the behavior becomes consistent.

**Comment by jaredhwilliams** [comment]
There is also a short delay between changing any configuration in the Product in Apigee Edge and having that change reflected and applied by the adapter.

OP, I'm not sure if this is your issue, but please watch out for these timing issues as they can bite you if you start attempting to test immediately after making a change (i.e. before the change has had time to fully propagate).

**Comment by theganyo** [comment]
Yes, by default, the polling refresh rate for product information from Apigee is every 2 minutes. This could be adjusted in config or you could [turn on debug level logging](https://github.com/apigee/istio-mixer-adapter/wiki/Mixer-logging-and-metrics) for the adapter and watch for lines like this to indicate that the product information has been refreshed:
```
2018-08-31T16:58:05.660590Z	debug	adapters	retrieving products from: https://theganyo1-eval-test.apigee.net/istio-auth/products	{"adapter": "apigee-handler.apigee.istio-system"}
2018-08-31T16:58:06.103321Z	debug	adapters	retrieved 3 products, kept 2	{"adapter": "apigee-handler.apigee.istio-system"}
```

**Comment by theganyo** [comment]
@aswinsegu Is it possible that your issue was caused by config latency as described above? Thanks.

**Comment by aswinsegu** [comment]
@theganyo The issue is not due to latency, as i had given enough time around 10 mins for each behaviour to act up on. Any how, had to clean the cluster and now its working from scratch up.

But need to see why that happened at all

**Comment by theganyo** [comment]
I cannot recreate this issue directly, but I believe this problem is likely due to a timing issue in the product dispatch that I've since fixed. I've also reviewed the code and don't see any other culprits. I'm going to close this, but I'll be keeping an eye out for any potential reasons why false rejections could occur. If it happens again, please also try to capture the mixer logs at DEBUG level. See: https://github.com/apigee/istio-mixer-adapter/wiki/Mixer-logging-and-metrics.

**Comment by AshishThakur** [comment]
Still see this error. The bug which you have mentioned has that been released in 1.0.4?
